Add Kyber hybrid curves to libssl

bbutch opened the pull request:
### Issues:
None

### Description of changes: 
This change adds kyber512-based hybrid handshake negotiation functionality to libssl. (The kyber crypto code was added previously to libcrypto.) The bulk of the functionality added in this PR is `PQHybridKeyShare` in `ssl_key_share.cc`. `PQHybridKeyShare` implements the handshake according to [this draft spec](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design). This implementation will interoperate with s2n-tls and liboqs. 

### Call-outs:
* This PR has been opened against the `integrate-pq` branch
* The main libssl tests rely on Golang's TLS implementation. We can't use that to test the kyber groups because Go doesn't include them.

### Testing:
* Updated `run_s2n_integration.sh` to test aws-lc/libssl handshakes against s2n-tls

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+
+  bool Accept(CBB *out_public_key, Array<uint8_t> *out_secret,
+              uint8_t *out_alert, Span<const uint8_t> peer_key) override {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
```
setting the alert here means the function will always return `SSL_AD_INTERNAL_ERROR` in `out_alert`, even on success. I don't think we should do this.

bbutch replied:
Right, I was just trying to keep it consistent with what is already done in the other key shares
* https://github.com/awslabs/aws-lc/blob/f6614260186c1b437ae54406fcb9b7f0736b154b/ssl/ssl_key_share.cc#L79
* https://github.com/awslabs/aws-lc/blob/f6614260186c1b437ae54406fcb9b7f0736b154b/ssl/ssl_key_share.cc#L161

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    *out_alert = SSL_AD_INTERNAL_ERROR;
```
setting the alert here means the function will always return `SSL_AD_INTERNAL_ERROR` in `out_alert`, even on success. I don't think we should do this.

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+    uint16_t ec_public_key_length;
+    Array<uint8_t> ec_secret;
+
+    if (!get_ec_public_key_length(&ec_public_key_length, ec_nid_)) {
```
the same comment as for the previous function: should we return more specific error codes?

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+  int pq_nid_;
+  uint16_t group_id_;
+  UniquePtr<SSLKeyShare> ec_key_share_;
+  EVP_PQ_KEM_CTX *pq_kem_ctx_;
```
do you think it's worth it to implement `PQKeyShare`class similarly to `SSLKeyShare` and `CECPQ2KeyShare` and then in `PQHybridKeyShare` just use `PQKeyShare` and `SSLKeyShare`?

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+    {NID_SECP256R1_KYBER512, SSL_CURVE_SECP256R1_KYBER512, "P-256_Kyber512", "prime256v1_kyber512"},
+};
+
+CONSTEXPR_ARRAY PQGroup kPQGroups[] = {
```
I think `PQGroups` is misleading because we have a mix of purely classical, purely PQ, and hybrid groups. It would probably be the best if we called them something with Hybrid.

dkostic commented on ssl/ssl_key_share.cc:
```diff
@@ -386,6 +583,23 @@ bool ssl_name_to_group_id(uint16_t *out_group_id, const char *name, size_t len)
+    return false;
+  }
+  const size_t field_len = BN_num_bytes(&ec_group->field);
+  *out_public_key_length = 1 + (2 * field_len);
```
why `+1`? it'd be nice to add a comment explaining why is the length computed in this way.

dkostic commented on ssl/ssl_test.cc:
```diff
@@ -422,6 +422,13 @@ static const CurveTest kCurveTests[] = {
       SSL_CURVE_X25519,
     },
   },
+  {
```
we don't test `P-256_Kyber512`?

bbutch commented on ssl/internal.h:
```diff
@@ -1118,12 +1118,27 @@ class SSLKeyShare {
 
+// PQGroup and PQGroups() are small helpers to encapsulate and keep track of
+// the component members of the hybrid named group.
+struct PQGroup {
```
I guess I'm being lax with the naming, since in my mind it's sort of implied that (for the foreseeable future) any "PQ Group" will be inherently hybrid. I can be more explicit in the naming with something like `PQHybridGroups`.

dkostic replied:
yeah, I think it'd be better to call them `PQHybridGroup`

bbutch commented on ssl/internal.h:
```diff
@@ -1118,12 +1118,27 @@ class SSLKeyShare {
+// the component members of the hybrid named group.
+struct PQGroup {
+  uint16_t group_id;
+  int nid;
```
The nid of the overall `PQHybridGroup`. I'll name it more explicitly.

bbutch commented on ssl/internal.h:
```diff
@@ -1118,12 +1118,27 @@ class SSLKeyShare {
+struct PQGroup {
+  uint16_t group_id;
+  int nid;
+  int ec_nid;
```
A general hybrid group isn't technically restricted to anything (it _could_ be EC + EC, EC + RSA, PQ + PQ, etc). But every PQ-hybrid group that I've ever seen defined for either TLS 1.2 or 1.3 has been PQ + EC. [This](https://github.com/open-quantum-safe/openssl/blob/474baededfff0b63668427dbe15841604f665949/oqs-template/oqs-kem-info.md) is the de facto group definition list for TLS 1.3.

bbutch commented on ssl/s3_both.cc:
```diff
@@ -666,7 +666,7 @@ class CipherScorer {
   CipherScorer(uint16_t group_id)
       : aes_is_fine_(EVP_has_aes_hardware()),
-        security_128_is_fine_(group_id != SSL_CURVE_CECPQ2) {}
+        security_128_is_fine_(!is_post_quantum_group(group_id)) {}
```
From what I can tell, it's used to evaluate cipher suites to find the minimum acceptable security for both client and server. It's relevant here, because, presumably, if you are using a PQ key exchange, you are concerned with a quantum adversary and so using AES-128 is not acceptable.

bbutch commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
 };
 
+// The hybrid key exchange implemented here is slightly different from the
+// key exchange used for CECPQ2. PQHybridKeyShare prepends 2 bytes to indicate
```
I'll clarify this comment.

* The structure of the bytes on the wire is defined in the ietf draft. I'll point to the section.
* The structure of the key share is:
IANA ID (2 bytes)
|| total share size (2 bytes)
|| size of ECC key share (2 bytes)
|| ECC key share (variable bytes)
|| size of PQ key share (2 bytes)
|| PQ key share (variable bytes)
* As far as I know, it's enough for every key exchange that is under consideration for standardization by NIST.

bbutch commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+    }
+
+    // Verify all lengths
+    if (peer_key.size() != ec_public_key_length + pq_kem_ctx_->kem->public_key_length + (2 * sizeof(uint16_t))) {
```
Sure. (It's 2 byte prefix for each public key.)

bbutch commented on ssl/ssl_key_share.cc:
```diff
@@ -290,13 +293,197 @@ class CECPQ2KeyShare : public SSLKeyShare {
+
+    // Verify all lengths
+    if (peer_key.size() != ec_public_key_length + pq_kem_ctx_->kem->public_key_length + (2 * sizeof(uint16_t))) {
+      *out_alert = SSL_AD_DECODE_ERROR;
```
I _think_ these error codes are just for internal use, not SSL error codes that get transmitted. But, either way, the convention in the other key shares is to set `SSL_AD_DECODE_ERROR` if the size verification fails e.g.

https://github.com/awslabs/aws-lc/blob/f6614260186c1b437ae54406fcb9b7f0736b154b/ssl/ssl_key_share.cc#L171

bbutch commented:
Updated with some significant reworking. Notably:

* abstracted the hybrid group concept into it's own HybridKeyShare, which supports any generic component groups (not just EC+PQ). 
* updated the handshake logic to conform to https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design

I'm converting this back to draft for now because I have to rework the tests. This implementation is no longer compatible with s2n, since s2n's hybrid handshake was implemented based on a previous version of the IETF draft. This implementation should be compatible with liboqs. I need to do some work to update/create tests accordingly.
